_posts/2022-06-07-RealWorldCTF-digging_into_kernel.md

```diff
@@ -0,0 +1,1128 @@
+
+We look at the value of the `user_data`: 0x10 = dq(0x8)+dd(0x4)+dd(0x4)
+
+````
+copy_from_user(&user_data, value, 0x10LL);
+````
+
+In other words, the `user_data` itself is a structure with `three offsets`  or   ` variables`.
+
+When the equation `cmd == 0x66666` holds, we can write the data of `user_data   ` of  length  `len`  to `buf+index`.
+
+```
+  if ( cmd == 0x6666666 )
+  {
+    if ( buf && len <= 0x50 && index <= 0x70 )
+    {
+      copy_from_user((char *)buf + (int)index, user_data, (int)len);
+      return 0LL;
+    }
+  }
+```
+
+When the equation `cmd == 0x111111` holds ,we can use the `kmem_cache_alloc` to allocate `0xc0` space to `s` from the `0xcc0 ` free block.
+
+```
+if ( cmd == 0x1111111 )
+        buf = (void *)kmem_cache_alloc(s, 0xCC0LL);
+      return 0LL;
+```
+
+
+
+When the equation `cmd == 0x777777` holds,we can use `copy_to_user` 
+
+to read data from `buf+index`offset address into `user_data`.
+
+```
+if ( buf && len <= 0x50 && index <= 0x70 )
+    {
+      copy_to_user(user_data, (char *)buf + (int)index);
+      return 0LL;
+    }
+```
+
+#### Trick
+
+https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/sched.h#L878
+
+In the kernel, each process is allocated a `task_struct` structure containing `process information`, `structure information`, and a `cred structure`.
+
+```
+	/* Empty if CONFIG_POSIX_CPUTIMERS=n */
+	struct posix_cputimers		posix_cputimers;
+
+	/* Process credentials: */
+
+	/* Tracer's credentials at attach: */
+	const struct cred __rcu		*ptracer_cred;
+
+	/* Objective and real subjective task credentials (COW): */
+	const struct cred __rcu		*real_cred;
+
+	/* Effective (overridable) subjective task credentials (COW): */
+	const struct cred __rcu		*cred;
+```
+
+The cred structure contains the `permission information` of each process,such as `uid`, `gid`,`suid`, etc.
+
+https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/cred.h#L111
+
+```
+struct cred {
+	atomic_t	usage;
+#ifdef CONFIG_DEBUG_CREDENTIALS
+	atomic_t	subscribers;	/* number of processes subscribed */
+	void		*put_addr;
+	unsigned	magic;
+#define CRED_MAGIC	0x43736564
+#define CRED_MAGIC_DEAD	0x44656144
+#endif
+	kuid_t		uid;		/* real UID of the task */
+	kgid_t		gid;		/* real GID of the task */
+	kuid_t		suid;		/* saved UID of the task */
+	kgid_t		sgid;		/* saved GID of the task */
+	kuid_t		euid;		/* effective UID of the task */
+	kgid_t		egid;		/* effective GID of the task */
+	kuid_t		fsuid;		/* UID for VFS ops */
+	kgid_t		fsgid;		/* GID for VFS ops */
+	unsigned	securebits;	/* SUID-less security management */
+	kernel_cap_t	cap_inheritable; /* caps our children can inherit */
+	kernel_cap_t	cap_permitted;	/* caps we're permitted */
+	kernel_cap_t	cap_effective;	/* caps we can actually use */
+	kernel_cap_t	cap_bset;	/* capability bounding set */
+	kernel_cap_t	cap_ambient;	/* Ambient capability set */
+	……
+
+	extern struct cred *prepare_creds(void);
+	……
+```
+
+Analysis of `prepare_creds`:

```

Comment:
You need to add analysis of the function [cred_init](https://elixir.bootlin.com/linux/latest/source/kernel/cred.c#L689). This helps you verify the length of this dedicated slab.

Reply:
Thank you very much for your correction, the analysis of the cred_init function has been added.

Reply:
Thank you very much for your corrections, which have added analysis of cred_init functions.